### PR TITLE
TN-1246 - always include iteration & recurs in QB calculations

### PR DESCRIPTION
**Pull request by pbugni**
Modified several functions used in calculating assessment status on QuestionnaireBanks.  The QuestionnaireBank alone lacks details needed for recurring QBs - therefore the named tuple QBD is now being used to capture vital details.  (QBD was in use previously, but not consistently)

### CODE REVIEW

**Comment by pbugni** [comment]
I'll note, when coupled with @mcjustin 's PR that includes the missing communication requests, that process is also working as expected:
```

  | "BEGIN trace for communicate on user 38"
-- | --
1 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
2 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
3 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
4 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
5 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
6 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
7 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
8 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
9 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
10 | "most_recent found QBD(relative_start=datetime.datetime(2018, 6, 4, 16, 0), iteration=1, recur=<Recur 1>, questionnaire_bank=<QuestionnaireBank 7>)"
11 | "process QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring; iteration 1"
12 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
13 | "trigger_date = 2017-09-18 16:00:00"
14 | "computed start 2018-06-04 16:00:00 for this questionnaire_bank"
15 | "computed expiry 2018-07-03 16:00:00 for this questionnaire_bank"
16 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"days\": 0} iteration 0"
17 | "iteration mismatch, request doesn't apply"
18 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 2} iteration 0"
19 | "iteration mismatch, request doesn't apply"
20 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 3} iteration 0"
21 | "iteration mismatch, request doesn't apply"
22 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 4} iteration 0"
23 | "iteration mismatch, request doesn't apply"
24 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 1} iteration 0"
25 | "iteration mismatch, request doesn't apply"
26 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"days\": 0} iteration 1"
27 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
28 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
29 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
30 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
31 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
32 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
33 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
34 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
35 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
36 | "most_recent found QBD(relative_start=datetime.datetime(2018, 6, 4, 16, 0), iteration=1, recur=<Recur 1>, questionnaire_bank=<QuestionnaireBank 7>)"
37 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
38 | "QuestionnaireBank status for QBD(relative_start=datetime.datetime(2018, 6, 4, 16, 0), iteration=1, recur=<Recur 1>, questionnaire_bank=<QuestionnaireBank 7>):"
39 | "  eortc:{'status': 'Due', 'by_date': datetime.datetime(2018, 6, 25, 16, 0)}"
40 | "  ironmisc:{'status': 'Due', 'by_date': datetime.datetime(2018, 6, 25, 16, 0)}"
41 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
42 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
43 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
44 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
45 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
46 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
47 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
48 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
49 | "found valid_consent with trigger_date 2017-09-18 16:00:00"
50 | "most_recent found QBD(relative_start=datetime.datetime(2018, 6, 4, 16, 0), iteration=1, recur=<Recur 1>, questionnaire_bank=<QuestionnaireBank 7>)"
51 | "found unfinished work"
52 | "notifiy_date 2018-06-04 16:00:00 has passed - add communication"
53 | "added Communication for user 38 of IRONMAN Recurring \| 9 Month Reminder (1)"
54 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 1} iteration 1"
55 | "found unfinished work"
56 | "notifiy_date 2018-06-11 16:00:00 has passed - add communication"
57 | "added Communication for user 38 of IRONMAN Recurring \| 9 Month Reminder (2)"
58 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 2} iteration 1"
59 | "found unfinished work"
60 | "notify_date 2018-06-18 16:00:00 hasn't yet come to pass, doesn't apply"
61 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 3} iteration 1"
62 | "found unfinished work"
63 | "notify_date 2018-06-25 16:00:00 hasn't yet come to pass, doesn't apply"
64 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 4} iteration 1"
65 | "found unfinished work"
66 | "notify_date 2018-07-02 16:00:00 hasn't yet come to pass, doesn't apply"
67 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"days\": 0} iteration 2"
68 | "iteration mismatch, request doesn't apply"
69 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 1} iteration 2"
70 | "iteration mismatch, request doesn't apply"
71 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 2} iteration 2"
72 | "iteration mismatch, request doesn't apply"
73 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 3} iteration 2"
74 | "iteration mismatch, request doesn't apply"
75 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 4} iteration 2"
76 | "iteration mismatch, request doesn't apply"
77 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"days\": 0} iteration 3"
78 | "iteration mismatch, request doesn't apply"
79 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 1} iteration 3"
80 | "iteration mismatch, request doesn't apply"
81 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 2} iteration 3"
82 | "iteration mismatch, request doesn't apply"
83 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 3} iteration 3"
84 | "iteration mismatch, request doesn't apply"
85 | "process eligable CommunicationRequest for QuestionnaireBank 7 IRONMAN_recurring_3mo_pattern recurring on {\"weeks\": 4} iteration 3"
86 | "iteration mismatch, request doesn't apply"
87 | "multiple (2) communications generated for single qb - mark all but one 'suspended'"
88 | "Sent 0 messages to pbugni+38@gmail.com after forced update"

```

File: tests/test_questionnaire_bank.py [hunk 1]
```diff
@@ -551,6 +551,18 @@ def test_visit_6mo(self):
         qbd_i2 = qbd._replace(iteration=1)
         self.assertEqual("Month 18", visit_name(qbd_i2))
 
+    def test_visit_9mo(self):
+        crv = self.setup_qbs()
+        backdate, nowish = associative_backdate(
+            now=now, backdate=relativedelta(months=9))
```
Comment (ivan-c) [inline tests/test_questionnaire_bank.py]: Don't like the non-determinism, but out of scope for this PR

Comment (pbugni) [reply tests/test_questionnaire_bank.py]: why not?  this is testing our system far better by moving through the days as the tests run on a nearly constant basis.  if we hard coded dates, it'd only test that one set.

Comment (ivan-c) [reply tests/test_questionnaire_bank.py]: It's true we do get more effective coverage, but we might be able to do that with a parameterized test run across a large number of days

Some [strong words from Martin Fowler](https://martinfowler.com/articles/nonDeterminism.html), still need to finish reading...

**Pull request merged**